edb/server/tenant.py

```diff
@@ -97,3 +139,112 @@ async def _pg_connect(self, dbname: str) -> pgcon.PGConnection:
     async def _pg_disconnect(self, conn: pgcon.PGConnection) -> None:
         metrics.current_backend_connections.dec()
         conn.terminate()
+
+    @contextlib.asynccontextmanager
+    async def _use_sys_pgcon(self) -> AsyncGenerator[pgcon.PGConnection, None]:
+        if not self._server._initing and not self._server._serving:
+            raise RuntimeError("EdgeDB server is not serving.")
+
+        await self._sys_pgcon_waiter.acquire()
+
+        if not self._server._initing and not self._server._serving:
+            self._sys_pgcon_waiter.release()
+            raise RuntimeError("EdgeDB server is not serving.")
+
+        if self.__sys_pgcon is None or not self.__sys_pgcon.is_healthy():
+            conn, self.__sys_pgcon = self.__sys_pgcon, None
+            if conn is not None:
+                self._sys_pgcon_ready_evt.clear()
+                conn.abort()
+            # We depend on the reconnect on connection_lost() of __sys_pgcon
+            await self._sys_pgcon_ready_evt.wait()
+            if self.__sys_pgcon is None:
+                self._sys_pgcon_waiter.release()
+                raise RuntimeError("Cannot acquire pgcon to the system DB.")
+
+        try:
+            yield self.__sys_pgcon
+        finally:
+            self._sys_pgcon_waiter.release()
+
+    def on_sys_pgcon_connection_lost(self, exc):
```

Comment:
Missing typing is added in a future commit https://github.com/edgedb/edgedb/pull/5742/commits/ea82cdf68269ed2b0ca64aa394f1d719a07bb116